packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,534 @@
    *
-   * @type {Object}
+   * @private
+   * @returns {boolean} true if there is a websocket (may be not open or not authenticated)
    */
-  _events = {}
+  hasWebSocket() {
+    return !!this.websocket
+  }
 
   /**
-   * Constructor of CozyRealtime:
-   * - Save cozyClient
-   * - create socket
-   * - listen cozyClient events
-   * - unsubscribeAll if window unload
+   * Is the websocket opened?
    *
-   * @constructor
-   * @param {CozyClient} client  A cozy client
-   * @param {CozyClient} cozyClient  A cozy client [DEPRECATED]
+   * @private
+   * @returns {boolean} true if opened, but may not be authenticated yet
    */
-  constructor({ cozyClient, client }) {
-    if (cozyClient) {
-      console.warn(
-        'Realtime: options.cozyClient is deprecated, please use options.client'
-      )
-    }
-    this._cozyClient = client || cozyClient
-    if (!this._cozyClient) {
-      throw new Error(
-        'Realtime must be initialized with a client. Ex: new Realtime({ client })'
-      )
-    }
-
-    this._updateAuthentication = this._updateAuthentication.bind(this)
-    this.unsubscribeAll = this.unsubscribeAll.bind(this)
-    this._receiveMessage = this._receiveMessage.bind(this)
-    this._receiveError = this._receiveError.bind(this)
-    this._resubscribe = this._resubscribe.bind(this)
-    this._beforeUnload = this._beforeUnload.bind(this)
-    this._resetSocket = this._resetSocket.bind(this)
+  isWebSocketOpen() {
+    return !!(
+      this.hasWebSocket() && this.websocket.readyState === WebSocket.OPEN
+    )
+  }
 
-    this._createSocket()
+  /**
+   * Is the websocket ready and authenticated?
+   *
+   * @private
+   * @returns {boolean} true if opened and authenticated
+   */
+  isWebSocketAuthenticated() {
+    return !!(this.isWebSocketOpen() && this.websocket.authenticated)
+  }
 
-    this._cozyClient.on('login', this._updateAuthentication)
-    this._cozyClient.on('tokenRefreshed', this._updateAuthentication)
-    this._cozyClient.on('logout', this.unsubscribeAll)
+  /* * * * * * * * * * */
+  /* * SUBSCRIPTIONS * */
+  /* * * * * * * * * * */
 
-    if (inBrowser) {
-      global.addEventListener('beforeunload', this._beforeUnload)
-      global.addEventListener('online', this._resubscribe)
-      global.removeEventListener('offline', this._resetSocket)
+  /**
+   * Subscribe to an event, type, id
+   *
+   * @param {EventName} event
+   * @param {string} type
+   * @param {string|undefined|function} id - (or handler is `handler` is undefined)
+   * @param {function|undefined} handler - (not used if `id` is a function)
+   */
+  subscribe(event, type, id, handler) {
+    this.start() //  start the realtime in case it was stopped
+    const sub = this.normalizeSubscription(event, type, id, handler)
+    const has = this.subscriptions.hasSameTypeAndId(sub)
+    this.subscriptions.add(sub)
+    // send to the server if there wasn't any with that type & id before
+    if (!has && this.isWebSocketAuthenticated()) {
```

Comment:
What if the websocket is not authenticated? Should we store the `subscriptions` we want to make and send them when the socket is authenticated?

Reply:
We always store the subscriptions, it is the role of `this.subscriptions.add(sub)` two lines above. The subscriptions are always all sent after authentication and future opening are already covered. This is  a special cas for the current already opened websocket.

Reply:
In that case, where are we removing the subscription if we send the Subscription successfully?

Reply:
Seen together.